### ISSUE TITLE
Experiments with node-tar to use js-virtualfs and random access

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/Polykey#1]
The node-tar is the best tar library in the Node ecosystem.

However it was designed with 3 things:

1. Asynchronous IO
2. Streaming interface
3. Interacting with a real fs
4. Compatibility with loads of tar options.

Since the tar archives that we are manipulating in-memory, we do not need async IO, we don't need streaming since all creation and extraction will be synchronous (even though real tar is actually streaming), and we want to interact with our js-virutalfs. We also don't care about all the tar options and the pax spec. We just want simple ustar implementation of creation and extraction and random access.

We also need to integrate random-access into our tar archives. So to be exact, we are compatible with the tar spec, but we include at the end of file tar index based on the go-rat repository (https://github.com/mcuadros/go-rat).

I have forked node-tar and investigat[ed its source code and also the ustar spec.

The creation function is very abstracted based on custom stream implementation.

The real meat of it is in the `lib/write-entry.js` that utilises the `header.js`. These are the things that really write the bytestrings into the resulting tar archive for any given file entry.

Once we have ripped out the appropriate things, our own "tar" library will be much more simpler and geared towards to in-memory creation of archives. In fact the entire archive can just be a virtual tar, that can be serialised to disk. And represent a "memory mapping" of tar archives. Or our in our case, a virtual tar is the in-memory lazy unpacked representation of the packed tar archive (which can be on disk).

Note that we also don't care about compression. So we won't even have that. This is pure tar. Compression only is relevant for disk purposes. Like real disk. But our situation is all in-memory. But remember we are taking tar archives, being able to extract them into memory on random access.

Tar format: https://github.com/libarchive/libarchive/wiki/ManPageTar5

**Comment by CMCDragonkai** [comment]
If the top-level key node is an encrypted tar archive. This means on every mutation to the secrets, you have to pack everything up via tar serialisation and encrypt it again. This can be quite slow.

**Comment by CMCDragonkai** [comment]
Further problem is that you cannot efficiently splice into a file of different lengths. This would mean every change involves reserialising the entire key node if it is a single tar archive.

It doesn't seem that tar archives are a good medum for maintaining secrets and being able to perform random access and random updates. Instead a directory structure makes more sense. However the encryption becomes more complicated, since we have to encrypt each individual file now, but if we do, we don't get the benefit of version control as much as we want (or we only get minimal rollbacks but always to encrypted files).

Also if the tar files exist only in memory, we can build the index in memory instead of forcing an encoding of them onto disk. The point of a virtual fs was also to make use of existing tooling like node-tar. If we only change the fs part of it. We still get most of the node-tar functionality? But at the cost of a clunky API. We should also investigate if this is satisfactory.

**Comment by CMCDragonkai** [comment]
How often do secrets change? It is far more likely that secrets are being read and shared then they are being changed. So it's ok if we have slightly more expensive writes than reads. Still using tar archives any kind of write to become a global serialisation of the entire key node and that's pretty bad.

One way to get around this is that the top level key node is not a tar archive (which means it cannot be encrypted portably), instead it's expected to be a directory. As a directory, each secret repository is then a tar archive that is encrypted. That way writes to secret repositories does involve a reserialisation and then reencryption. But doesn't affect other key repositories.

**Comment by CMCDragonkai** [comment]
If we discard the tar concept, we lose a bit. But the main idea is that each secret repository is engaged in voluntary state replication with other keynode peers.

**Comment by CMCDragonkai** [comment]
That would mean you stop using git as the medium of replication, and instead use something like: https://github.com/mafintosh/hyperdrive Not sure how that will interact with encryption however.

https://github.com/mafintosh/hyperdrive/issues/190

However we lose git version control if we use this. The best way is to either build something that combines version control and p2p state replication automatically.

**Comment by CMCDragonkai** [comment]
The direction right now will be to have unencrypted directory for a keynode. And within this directory will be secret repositories.

Each secret repository is an encrypted tar archive. They will be decrypted into memory and unpacked for mutation. Before saving a secret repository onto disk. The contents will be created as a virtual tar archive, encrypted in memory, content hashed with a sha256sum (or multihash) which will be used as the name of the secret repository. Alternatively the name can be chosen by the user. The name will be important for choosing which secret repository to investigate.

This means the names are not bound to the secret repository globally. One keynode might call a secret repository A, while another may call it B.

**Comment by CMCDragonkai** [comment]
We are no longer using node-tar, instead we are using tar-stream, it's far more conducive to in-memory manipulation.

Since we changed from a global encrypted tar archive to a keynode directory, random access isn't needed as much. Also the main issue is that as we make changes to a secret repository, in order to persist those changes, they must be "reserialised" back into the disk. Reserialising and reencrypting a secret repository is a linear operation. Regardless of whether we implement random access into the tar archive. Basically even if there was a random access index in the tar archive, the best we can do is to inject an extra file block into the end of the tar stream, and move the index to end. While if we removed a file, or edited a file, that may change a block size, that will require re-indexing the entire tar archive. On top of that, because we have encryption, the whole structure is just not designed for random mutation.

If we really want random mutation, we have to investigate: https://en.wikipedia.org/wiki/Disk_encryption_theory

And then we would have to implement a special "filesystem" structure to have efficient secrets manipulation. It would have be some sort of database.

For now I'm going to push that to later in another issue, and the initial prototype will just use encrypted tar archives.
